### PR TITLE
refactor(error): reduce error type size

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR reduces size of the following error types:

| Error               | Before Size | After Size                  |
| ------------------- | ----------- | --------------------------- |
| StreamError         | 240         | 8                           |
| StreamExecutorError | 240         | 8 |
| ExprError           | 184         | 40                          |
| UdfError            | 184         | 32                          |
| RpcError            | 176         | 16                          |

by boxing the error body and offloading them to the heap.

The common pattern of this technique is:
```rust
pub struct Error {
    inner: Box<Inner>,
}
struct Inner {
    kind: ErrorKind,
    backtrace: Backtrace,
}
enum ErrorKind {
    Storage(StorageError),
    ...
}
```

This partially _resolves_ #11263.

## Checklist

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

### CODE REVIEW

File: src/expr/src/error.rs [hunk 1]
```diff
@@ -50,10 +48,13 @@ pub enum ExprError {
     DivisionByZero,
 
     #[error("Parse error: {0}")]
-    Parse(Cow<'static, str>),
+    Parse(Box<str>),
```
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/expr/src/error.rs:51]: Box takes 16 bytes, while Cow takes 32 bytes.

**Comment by wangrunji0408** [comment]
I'm considering whether the `ExprError` is also worth reducing to 8 bytes. Because it's used in scalar functions which are also performance sensitive.

**Review by xxchan — COMMENTED** [review]
🔥🥵
- merge StreamExecutorError and StreamError: not sure. What’s the reason? I generally prefer more fine-grained error types, if no much trouble introduced, because it made clear what’s possible errors for a function. 
- XxError -> Error. Not sure. Personally slightly prefer the former, but no strong opinions.

**Comment by wangrunji0408** [comment]
> merge StreamExecutorError and StreamError: not sure. What’s the reason? I generally prefer more fine-grained error types, if no much trouble introduced, because it made clear what’s possible errors for a function.

I found the main body of stream crate is the executors. Having two error types cause double memory usage in most cases. And I prefer one unified error type in one crate.

> XxError -> Error. Not sure. Personally slightly prefer the former, but no strong opinions.

In most 3rd party libraries and std library, their error types are named `Error`. 🤔

**Comment by BugenZhao** [comment]
> And I prefer one unified error type in one crate.

Personally, I prefer to keep the granularity as fine as possible. 😰 Currently the difference between `StreamError` and `StreamExecutorError` is kind of clear. The former is for controllers and managers, and the latter is for executor computations.

Similar ideas are applied to expression as well. -> #10764

**Comment by TennyZhuang** [comment]
I'd prefer split Error instead of merge Error.

**Comment by wangrunji0408** [comment]
Reverted merging and renaming. PTAL again. 🤡

File: src/rpc_client/src/error.rs [hunk 2]
```diff
@@ -21,19 +21,33 @@ pub type Result<T> = std::result::Result<T, RpcError>;
 #[derive(Error, Debug)]
 pub enum RpcError {
     #[error("Transport error: {0}")]
-    TransportError(#[from] tonic::transport::Error),
+    TransportError(Box<tonic::transport::Error>),
```
Comment (BugenZhao) [inline src/rpc_client/src/error.rs:24]: So `#[from]` does not work with a wrapper?

Comment (BugenZhao) [reply src/rpc_client/src/error.rs:24]: BTW, `#[from]` also helps us to implement `source` and `provide` (like backtrace). :(

File: src/udf/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use arrow_flight::error::FlightError;
+
+/// A specialized `Result` type for UDF operations.
+pub type Result<T> = std::result::Result<T, Error>;
```
Comment (BugenZhao) [inline src/udf/src/error.rs:18]: nit: I find in this way we can make it more flexible:

https://github.com/dtolnay/anyhow/commit/fdd60501d7e93eab70fdb584588ef6199b1c61b3

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM

File: src/rpc_client/src/error.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/rpc_client/src/error.rs:24]: Right... How can I manually implement these?

Comment (wangrunji0408) [reply src/rpc_client/src/error.rs:24]: I see. I can keep the `#[from]` over box and forward `impl From<InnerError>` to from box.

**Review by BugenZhao — APPROVED** [review]
LGTM! Really appreciate.

**Comment by MrCroxx** [comment]
Nice! Looking forward to seeing the performance!

**Pull request merged**